### ISSUE TITLE
Redesign ValueProviderResult + Collection Model Binders

### ISSUE THREAD

**Issue by rynowak** [aspnet/Mvc#2268]
From API review:

Currently a model bind for a collection can go down a number of paths depending on what a value provider returns.
- A result with a collection as it's value is handled by `TypeMatchModelBinder`
- A collection of results are handled by collection model binders

This needs more design but we want to eliminate one of these code paths

**Comment by danroth27** [comment]
@loudej This has limited customer impact, so it seems like low priority.

**Comment by lodejard** [comment]
Medium-ish impact in the sense it's a relatively common subsystem to extend, and is extremely intricate as it stands.

**Comment by anfomin** [comment]
Current `CollectionModelBinder` can't properly bind collections with dynamic indexes. It returns invalid results in 2 cases.

**1.** I've class with collection property:
```C#
class RootObject
{
	public IList<AnotherObject> Collection { get; set; }
	// other code goes here
}

class AnotherObject
{
	public string Name { get; set; }
}
```

Binding form data:
```
Collection.index = key1,key2
Collection[key1].Name = name1
Collection[key2].Name = name2
```

If I try to bind `RootObject` then `Collection` property is bound successfully, but ModelState contains unvalidated states for `Collection[key1].Name` and `Collection[key2].Name`:
```C#
var model = new RoomObject();
TryUpdateModel(model);
/// ModelState.IsValid returns false - wrong value.
```


**2.** I've the same classes. If I try to bind collection of `AnotherObject` directly with prefix, result collection is still empty:
```C#
var collection = new List<AnotherObject>();
TryUpdateModel(collection, "Collection");
// collection is still empty, even if there is data
```

**Comment by dougbu** [comment]
@anfomin at least your first point is about validation, not model binding, and both may relate to bugs we recently fixed.  please:
- file new (individual) issues rather than adding comments to a design discussion
- let us know what MVC version you are testing

**Comment by rynowak** [comment]
This was addressed by 6d365e9a

Collections now have one code path for model binding.
